technical-docs/K8SSANDRA-OPERATOR-DESIGN.md

```diff
@@ -0,0 +1,410 @@
+
+Multi-cluster is out of scope, but I want to point out that I think this is where we introduce it. Consider the following example:
+
+```
+apiVersion: k8ssandra.io/v1alpha1
+kind: CassandraCluster
+metadata:
+  name: k8ssandra
+spec:
+  clusterName: k8ssandra
+  auth:
+    enabled: true
+  datacenters:
+  - name: dc1
+    size: 3
+  - name: dc2
+    k8s:
+      kubeconfigSecret: my-other-cluster
+      namespace: dev
+    size: 3
+```
+
+`kubeconfigSecret` specifies the name of a secret that contains the kubeconfig file needed to access a remote cluster. 
+
+`namespace` indicates in which namespace the DC should be created.
+
+### CassandraDatacenter
+
+This is defined in the [cass-operator](https://github.com/k8ssandra/cass-operator) repo. The controller runs in its own Deployment.
+
+### Reaper
+
+```
+apiVersion k8ssandra.io/v1alpha1
+kind: Reaper
+metadata:
+  name: k8ssandra
+spec:
+  serverConfig:
+    storageType: cassandra
+    jmxUserSecretName: k8ssandra
+    cassandraBackend:
+      cassandraDatacenter
+        name: dc1
+      replication:
+        networkTopology:
+          dc1: 3
+```
+
+This is defined in the [reaper-operator](https://github.com/k8ssandra/reaper-operator) repo. It is possible today to use reaper-operator on its own without k8ssandra. In that regard, living in its own repo makes sense. There is a good bit of overhead for each repo that we need to maintain. In practice, I do not think reaper-operator will be used a whole lot outside of k8ssandra. For these reasons I am in favor of consolidating with the k8ssandra-operator repo.
+
+From a technical standpoint, consolidation means that the reaper-operator controllers would run in the k8ssandra-operator binary. It would reduce overhead in that it would be able to use the same cache as other controllers.
+
+### Stargate
+
+This does not yet exist. Similar to reaper-operator, one could certainly make the case that a separate stargate-operator project makes sense. I am more on the fence about where this should live than I am with reaper-operator.
+
+```
+apiVersion: k8ssandra.io/v1alpha1
+kind: Stargate
+metadata:
+  name: k8ssandra
+spec:
+  replicas: 3
+  cassandraDatacenter:
+    name: dc1
+```
+
+### CassandraBackup / CassandraRestore
+
+These are defined in the [medusa-operator](https://github.com/k8ssandra/medusa-operator) repo. I propose consolidating this into the k8ssandra-operator repo.
+
+See this [GitHub discussion](https://github.com/k8ssandra/k8ssandra/discussions/273) more for ideas around backup/restore and CRDs.
+
+### MonitoringConfiguration
+
+```
+apiVersion: k8ssandra.io/v1alpha1
+kind: MonitoringConfiguration
+metadata:
+  name: k8ssandra
+spec:
+  prometheus:
+    # prometheus properties...
+  grafana:
+  # grafana properties...
+```
+
+With the Grafana license changes and discussion around Victoria Metrics, I think it is a good idea to encapsulate Prometheus and Grafana.
```

Comment:
+1 here, we'll also need to provide the ability for the control of the version of Grafana being deployed, so that people have the choice of pre-AGPL licensed versions.